Spider Rework

IndieanaJones opened the pull request:
## About The Pull Request

Closes #46322 (by getting rid of the special AI until we have something better in place)

This PR reworks giant spiders from xenobiology, changelings, and botany fame, mostly by modifying some of their capabilities to be more fun and balanced for both sides of the equation.

**WHAT WAS WRONG WITH SPIDERS**
- Generally got past an unstoppable point before anyone knew they existed.  Since spiders required humans or monkeys to create eggs and the traitor type that makes them usually has plenty of monkeys, spiders would usually have insane numbers before the crew even knew they existed.  Since a bunch of spiderlings could be made from one egg let alone a room-full and the nature of ventcrawling spiderlings, the spiders were everywhere as well.  Once the spiders grew, the crew faced an almost always losing battle against a never-ending tide.
- Numbers weren't reflective of spiders' status as a swarming group antag.  For reference, a basic red-eyed spider had 200 health, which is on par with an unarmored spaceman plus factor in that spiders don't go into crit, it might as well be double.  This allowed spiders to curbstomp easily with little thought.
- Had issues with metagaming.  This is something more reflected in terms of admins and whatnot, but there was an issue with spider players just taking a spider and using meta info to return to the same spot (which was incredibly fast due to ventcrawl) and continue attacking the same people over and over.  This led to the creation of some unsavory PRs which would have destroyed spiders in general, which is also not good.
- Non-player controlled spiders were essentially useless.  Due to their broken AI, they would never do anything interesting other than attack someone every so often.  This generally made changeling's spider power sort of a waste, as it just made hunters who were fairly easy to dispatch since there was no player piloting them.  The spider event was also a waste because of this as well, being on parity with xenomorph rarity yet being a much worse event in terms of anyone enjoying it and general threat.  This also sort of ties in to how spiderlings would just become rooms full of spiders which didn't do anything until a ghosts possessed them one by one and had them do stuff.

**THE CHANGES**

Essentially, these changes aim to hone in on the interesting aspect of spiders being waves of monsters which keep coming until you destroy their source.  Spiders should be changed to be made weaker in general because its expected that as long as the supporting roles continue their jobs, you should be able to get back into it fairly quickly.  However, spiders shouldn't be able to choose to spawn from just anywhere in the station, but from centralized locations maintained by broodmothers, as for why you'll see later.

**ALL SPIDERS**
- Ventcrawling removed.  The ability to vent crawl is extremely powerful, especially for tanky mobs.  It also allowed for spiders to perform well in that metagaming speedrun issue mentioned earlier, since unlike other mobs with similar abilities, spiders didn't (and still don't) have a "down-time" for meta info such as location to become irrelevant.  This problem is further fixed in a change listed later.
- Object damage reduced from 60 to 30.  This is rather minor, but it makes spiders take longer to break some structures.
- No longer speak or understand common, but speak and understand buzzwords.
- Stamina attacks now affect spiders and will slow them down temporarily.

**RED-EYED SPIDERS**
- Health reduced from 200 to 80
- Damage buffed from 15-20 to 20-25
- No longer injects poison

These guys are our grunts, but they just used to be too tanky for a swarming mob.  These changes aim to make them more manageable but rewarded for the hits they do land more.  The poison nerf isn't much, since they barely did anything anyway.

**PURPLE-EYED SPIDERS**
- Health reduced from 100 to 50
- Speed minorly buffed from 0 to -0.1.  While these guys became much weaker physcially, they can now catch up to you, but you can use stamina draining attacks to slow the buggers down for a while.
Moving on with the other health reduction of the patch, purple-eyed spiders now show their worth amongst their red-eyed peers by being the faster but less damaging of the two, as compared to just straight up worse as they were before.

**GREEN-EYED SPIDERS**
- No longer plays a role in the reproductive cycle.  Can no longer succ or lay eggs.
- Now lays down web 4x faster than other spider subtypes
- Can now heal other spiders after a do_after of 2 seconds for 20 health.  Useful if a spider survives a scrape so you can keep them alive for the next one, an ability especially helpful for the special spider variants.
- Gets a medHUD to see the health of spiders and others

So yeah, nurses don't get to lay eggs anymore.  While that might seem bad right now, keep going and I'll just assure you for now I'm not killing spiders.  Nurses are now more of a utility role, placing down large amounts of webbing and healing other spiders should they be able to stop what they're doing to be assisted.

**TARANTULAS**
- Now moves at 1 speed while on web, and very slowly while not on web
- Can now lunge using Arcane's tackling system.  6 second cooldown, and my mob lets players do it too :C)
- Obj damage buffed from 60 (everyone else is 30 now) to 100

First off, I fixed the speed thing with tarantulas, since it didn't work at all before.  You might wonder why I gave them a lunge, and the reason is simple: I didn't think they were threatening enough.  I see tarantulas the tanks of the spiders, but their 7 speed just made them unable to threaten anyone but cripples.  Now, you'll have to keep an eye on a tarantula nearby when you're fighting the horde in case it intends to pounce you, as if it lands it will surely kill you with its massive power.

**VIPER**
- Damage buff from 1 to 5
- Venom injection decreased from 12 to 6
- Speed buffed from 0 to -0.5 (Mi-go speed)

Vipers are now more moved into the hyper-specialized role of a fast attacker, being able to quickly pursue prey and deliver does of venom while quickly dodging attacks.  While they only have 40 health, inability to stop them before they land enough blows will cripple even the strongest of defensive forces.

**BROODMOTHER**
Here's where we get interesting, and everything starts to make sense.
- Can now always lay eggs.  Eggs themselves have been changed.  Instead a timer until a bunch of spiderlings spawn, it is now a timer until a ghost can interact with the egg cluster.  Upon doing so, a ghost can now play as a spider, and these "free" eggs will always produce one of the 3 basic units of green, purple, and red, with the choice being random.  However...
- Draining HUMANS (not monkeys) allows the broodmother to lay a single enriched egg cluster, which is a separate ability.  The spider which emerges from this egg cluster will always be one of the three special classes (Tarantula, Viper, Broodmother).  So, this gives an incentive to hunt and bring humans back to the broodmother, who can then use them to make the specialized classes.  The only major issue I see with this is Genetics, but if it is problematic there are ways around it.
- Laying eggs now takes 10 seconds instead of 5

So with these changes, you can see how this will play out.  Broodmothers will work in their own little areas to provide enough eggs to allow the swarm to spawn, and then must either continue producing more eggs as the horde dies or using their command ability to try and steer the direction of the horde's efforts.  Everyone has a role to play, with red and purple being frontline combantants with purple having an easier time approaching targets and damaging them so reds can catch up and finish fights, and green can serve on both the back and front line assisting whoever needs it.  Tarantulas serve as a "boss" in a sense that they're very tanky and require specific attention from crew.  If the crew managed to overwhelm the spiders, they can kill the broodmother by making their way to her location and then destroying the remaining eggs.  With more broodmothers, with effort they can be spread out as to provide multiple angles of attack for spiders to utilize, which takes actual effort as opposed to spiderlings doing the work for you prior.  Speaking of them:

**SPIDERLINGS**
- Now only can be produced from botany and the abductor gland, also meaning outside of pre-spawns, xenobio, botany, and the abductor gland, AI spiders do not appear.

And that brings us into the last part of our talk here:

**CHANGELING'S SPIDER ABILITY**
- No longer vomits up changelings, but instead bloody eggs.  Bloody eggs work like any other set of eggs do, but instead produce a flesh spider as its only option of spawn (mind you, that also means ghost controlled).  Flesh spiders are stat equivalent to purple-eyed spiders or hunters but also have the following traits:
- Can place web in 0.7 the amount of time a normal spider can
- Can heal itself, but not other spiders
- Occasionally leaves blood on the floor

**OTHER CHANGES**
- The spider infestation event now creates two egg clusters which produce broodmothers on two sufficiently-dark xeno spawns as its event as opposed to spawning spiderlings in the vents.  Therefore, spiders are now a midround antag of themselves, technically.  The event rarity has been set on parity with blob.
- The above event has also been added as a midround spawn in dynamic mode
- The entirely of the giant_spider.dm file and spider eggs have been auto-doc'd
- Removed the broken AI for giant spiders.  They act as regular hostile simplemobs for now.
- All spider subtypes now directly stem from the base red-eyed spider instead of the special types stemming from the base types.  There's enough differences now (espeically between broodmother and nurse) that this seemed unwise to keep.

## Why It's Good For The Game
A lot of people enjoy spiders, myself included, because its a ghost role everyone who wants to get involved is allowed to do so.  Of course, while it can be fun for some, the balance made it hard for the crew to prevail in most situations and just made spiders a thing which forced a shuttle call always unless a xenobiologist really messed up.  Now, with these changes, it takes a little more effort from the spiders to wipe the crew while leaving it possible to do so, hopefully making the situation of fighting spiders more fun and dynamic overall.  Also, changeling ghost role seemed like a neat idea in my head, and I believe since the ability is gated behind 3 drains, it should be balanced enough to not upset anything.  Also makes the other times when spiders could show up more interesting by putting actual players in control of them.

## Changelog
:cl:
add: New spider subtype - Flesh Spider.  Unique to changeling.
balance: Changeling's spider ability now creates a bloody egg cluster, which produces a singular player-controlled flesh spider per cluster.
balance: Spiders have been rebalanced.
/:cl:

IndieanaJones committed:
Really revert commit

IndieanaJones committed:
Now Undo the simple change

IndieanaJones committed:
Undo My Stupid yet again

IndieanaJones committed:
Undo my stupid yet again

IndieanaJones committed:
No

IndieanaJones committed:
And for the last time

actions-user committed:
Automatic changelog compile [ci skip]

actions-user committed:
Automatic changelog compile [ci skip]

IndieanaJones committed:
Revert "Automatic changelog compile [ci skip]"

This reverts commit 48a63defa644ec9826ee5eb1736e5158d2a636d4.

IndieanaJones committed:
Revert "Automatic changelog compile [ci skip]"

This reverts commit cb68f4988edd9d841c4203eecee2067a49ce75c5.

IndieanaJones committed:
Revert "Revert "Automatic changelog compile [ci skip]""

This reverts commit e6e1f1ef53ee6611c334efcae6ca7c102edaf351.

IndieanaJones committed:
Revert "Revert "Automatic changelog compile [ci skip]""

This reverts commit 1782210527b9819772c0781122bec35b7979bf7d.

IndieanaJones committed:
Update changelog.html

actions-user committed:
Automatic changelog compile [ci skip]

IndieanaJones committed:
All the changes to the spider file

IndieanaJones committed:
Change the Eggs

IndieanaJones committed:
Adjust the Tarantula's Movement Modifier

IndieanaJones committed:
Update Australium Extract Reaction

IndieanaJones committed:
Update the crappy shapeshift spell (why would you want this)

IndieanaJones committed:
Update the Changeling's Spider Ability

IndieanaJones committed:
Add a Spider Language Holder

IndieanaJones committed:
Update Icons

IndieanaJones committed:
Add Midwife Eggs for Random Event

IndieanaJones committed:
Update Spider Event

IndieanaJones committed:
Add Spiders to Dynamic

IndieanaJones committed:
Prevent Runtime

necromanceranne commented:
Are the flesh spiders friendly or hostile to the ling?

IndieanaJones commented:
> Are the flesh spiders friendly or hostile to the ling?

The two parties can't understand each other, so I guess neutral really.  I could directly state what they should do to their parent if that's something wanted

lolman360 commented:
the burn increase, health decrease combined might be a little much, but I probably underestimate the amount of ghosts that'll be spawning in as these things
i'd like to see how it plays out ingame

necromanceranne commented:
> > Are the flesh spiders friendly or hostile to the ling?
> 
> The two parties can't understand each other, so I guess neutral really. I could directly state what they should do to their parent if that's something wanted

Iunno them being neutral and thus unreliable makes sure lings can't just amass a personal army somewhere. At least lings finally would have a neat way to keep people in the game as one of the most notorious round remover antags.

Winterous commented:
I love love love love this PR!
Few questions.
1. Is blob more or less common than xenos?
That's all my questions.

Now for suggestions.

1. I think the hunter spiders should maybe be able to ventcrawl. Making the horde need to smash their way through everything to get anywhere just seems too limited, and these guys are smaller and more agile than most of their peers; allowing them to ambush people in isolated rooms, run interference with healing or re-arming, deliver a few bites and run so they're forced to heal. But maybe reduce their object damage a lot so they don't just smash everything in a room and leave. Hell, maybe be unable to smash stuff at all.
2. The nurse spiders should probably get a med-hud, at least for other spiders, so they know who needs healing.
3. And also I think that it would be wise to have on-spawn prompts telling you how your 'class' works, just one or two lines.
4. https://github.com/tgstation/tgstation/blob/eb1a9d7a679a3241d5dffc4af9c2320a6c4c4fd1/code/modules/mob/living/simple_animal/hostile/giant_spider.dm#L382 - This should specify that it allows you to lay ENRICHED eggs.
5. I recommend changing their temperature damage, they take 1.5x burn damage, but they take the same damage from being too hot or too cold; maybe it should be 30-40 damage from hot, and 10-15 from cold? Flamethrowers should probably be quite dangerous to them.

Now on the topic of flesh spiders, and also their relationship with changelings.
If they are to have a friendly relationship with changelings they should have some visual identification of who the changelings are; but I don't think that's the right way to go anyway.
I think that a neutral relationship would be best, but they could probably use an OBJECTIVE, rather than just "I exist therefore I kill".
Perhaps if they cocoon human bodies they could lay flesh eggs, or get stronger and tougher, up to a limit, and potentially gain the tarantula's tackle ability after a point?
If you do that then their objective would just be "survive, consume, and grow / reproduce". Without that they don't really anything in the way of motivation, they might even decide to be chummy with the crew, which would just be weird.

ATH1909 commented:
I like this PR, but I feel like new!red-eyed spiders will be by far the weakest of all of the spider types, and also the least fun to play as. Sure, they can hit reasonably hard in melee, but they can't actually GET into melee range with anyone who isn't already done for, like cult juggernauts. Unlike cult juggernauts, however, they have no utility abilities (that other spiders don't already get/can do better than them), aren't damage sponges, don't have a(n admittedly shitty) ""ranged attack"", and can't deflect projectiles. Their movement speed looks like it'll be sluggish enough to make moving around the station a chore (if you've played a masterless, bot-less pAI and tried roaming around the station in holoform, you'll know what I mean), and it looks like their gameplay loop consists of "roam station looking for prey, watch helplessly as prey just runs away from you on sight because you're a giant fuck-off spider, cry, repeat". About the only possible use I could think of for them would be to be dragged around by a purple-eyed spider (or even a viper, if one's available) to give them a bit of extra oomph, but tbh, a second purple-eyed spider would probably be better in that situation.

Coordination with a viper could MAYBE allow red-eyed spiders to catch up to prey, but that relies on the nest getting a competent viper/properly getting off of the ground, and the red-eyed spiders not being intercepted by healthy people with guns. Oh yeah, that's another reason why new!red-eyed spiders suck: they have no escape options, so if someone shows up with a gun, they can't really do anything to avoid being gunned down (other than praying that the person with a gun is low on ammo or an awful shot)- they can't move fast enough to close to melee range, and they also can't move fast enough to run away.

I'm still a bit iffy on purple-eyed spiders as well, since they're also gonna have a hard time chasing down prey unless they're really good at reading people. They can at least TRY to, like, ambush or pincer attack people or something, though, or maybe do something with people getting momentarily caught by webs.

Maybe give green-eyed spiders the ability to lay down web walls (and/or web floor paneling?) or something? I feel like a single plasmaflood, random bloke with a flamethrower, random engiborg with a plasma canister, or breach to space will be able to fuck over an entire spider colony with no possible counterplay from the spiders other than "spread out more, I guess". Speaking of which, how are the broodmothers from the random spider event expected to get their first kill(s) so that they can start making spiders, especially if they started in maint on the opposite side of the station from Virology, Xenobiology, and Genetics? Maybe you should give broodmothers (and red-eyed spiders too?) their ability to ventcrawl back, so that they can at least choose where to set up without leaving an incredibly obvious trail of smashed doors behind them.

Finally, remember that vipers are a possible form for the wizard version of the Shapeshift spell. The previous main draw of the form was it being a ventcrawling form that didn't die in one hit to almost anything, unlike the mouse form. With these changes, it loses that niche. I mean, I guess it's an escape option now? You can transform and skitter away if you get hardstunned or staminacritted, I suppose? But if you're in a situation where need an escape option, you probably don't want to switch to an even MORE flimsy/fragile form...

IndieanaJones commented:
I can think of something for the red-eyed spiders to give them a little more edge or at least more interesting gameplay.  As I said before, I don't want to give ventcrawl back because of the metagaming concern expressed by the admins as ventcrawling was pretty much its main enabler.  Flesh spiders can get an objective and I can also set up a team antag datum for spiders to get listed off at the end of the round like others do.  I'll also consider counterplay to fire floods and something for wiz I guess.

> are the broodmothers from the random spider event expected to get their first kill(s) so that they can start making spiders, especially if they started in maint on the opposite side of the station from Virology, Xenobiology, and Genetics? 

They can make infinite regular eggs, only being constricted by time.  Those other spiders can go out and bring back bodies for the broodmother to make enriched eggs.

Dorsisdwarf commented:
I think that mass removal of ventcrawling along with punishing speed nerfs and extreme health nerfs (3 shots with a laser kills red, 2 shots with a laser weapon kills purple) to the generic spiders is a pretty devastating combination, especially since spiders will now be more limited in the number you can have at any one time...

Winterous commented:
Yeah what they're saying about the movespeed seems pretty solid; maybe reds should move at normal speed, and purples should be a little bit faster than normal. If they're gonna be so fragile they need to be able to move quickly.

Also about the plasma flood stuff; maybe they could just clog vents with webs, or cover them over with a solid mass of web, like how resin blocks vents.

And yeah thinking about it, spiders being vulnerable to fire is a meme, doesn't make any real sense.

Tlaltecuhtli commented:
you know what would be cool? if they get moved to carbon level so they can get stunned by shocks and actually set on fire, dismembering legs, bug spray smoke breathing, which is what makes shit fun and balanced

ATH1909 commented:
> you know what would be cool? if they get moved to carbon level ((for sentient ones)) so they can get stunned by shocks / batons and actually set on fire, dismembering legs, bug spray smoke breathing, which is what makes shit fun and balanced

Wouldn't that make them even less distinct design-wise from xenomorphs?

IndieanaJones commented:
> you know what would be cool? if they get moved to carbon level ((for sentient ones)) so they can get stunned by shocks / batons and actually set on fire, dismembering legs, bug spray smoke breathing, which is what makes shit fun and balanced

I'm sure that would be cool, but that massively increases the amount of maint to keep them in working order and makes them much more suseptible to breaking.  I'll just admit I don't have all the time in the world to dedicate to my work on /tg/, so I think its best they stay simple mobs.  Not to mention, being able to stun them and whatnot doesn't seem very fun for the spiders, and if I wanted any of these individual features, I could add them in my own merit.

ATH1909 commented:
Would it be possible to let ghosts choose which kind of spider they'll spawn as?

IndieanaJones commented:
Anyway, the changes I'll be looking to implement from feedback:
- Revert red-eyed spider speed nerf
- Make purple-eyes (hunter) spiders slightly faster than humans as opposed to slightly slower
- Revert burn damage change in lieu of the health nerfs being able to cover making the spiders easier to kill by themselves
- Make the flesh spiders explicitly hostile to any non-spider life, including changelings
- See what I can do to add to give green-eyed (nurse) spiders med hud, should be easy
- Give blurbs on how to play each spider on spawn via Login()
- Mention enriched in the one ability description
- I guess make a wizard subtype of the viper with extra bonuses

Though, I was also thinking about this one, since I'd revert the extra burn damage and speed nerfs, I could make them slow down if stamina damage is dealt to them (they'd never go into stun though).  It would allow a higher diversity of weapons to be used against them as opposed to other simple mobs, it could prove interesting.  Though of course, the Tarantula would probably be very resistant to slowdown since they're already very slow to begin with.

Also, about the issue of burning them using the vents, could they not destroy them to prevent that?

IndieanaJones commented:
> Would it be possible to let ghosts choose which kind of spider they'll spawn as?

Yes, but can you make an argument for it?

Ratvar commented:
Player will have a specific spider type it likes playing as more, or knows how to play at least. Letting them chose will help them enjoy spiders. Player that wants to actually help spiders win will pick type that is more suitable, like going green-eyed spider if webs are really needed, making spiders more threatening then RNG deigns them to be, helping round progression.

Winterous commented:
Yeah people would totally re-roll their spiders.
And stamina damage actually doing something against simple mobs sounds very interesting indeed! Finally those newbie seccies disabling the space dragon WON'T be wasting effort!

evilcyborgmrt commented:
Spiders should be able to vent crawl

IndieanaJones commented:
> And stamina damage actually doing something against simple mobs sounds very interesting indeed! Finally those newbie seccies disabling the space dragon WON'T be wasting effort!

Ah, well, for the sake of this PR, it'd be limited to spiders.  If someone else wants to make it a trait of all simplemobs, then they'd be free to do so.

RuneQuester commented:
I like it

nianjiilical commented:
while you're working on spiders, would it at all be possible to give the three basic types proper names, since its not exactly clear which do what to new players

also this might be out of the scope of this pr but itd be really cool if sergeant aranus was his own subtype with some sort of security-focused ability (web cuffs?) since sentiencing him to work as a security spider is fun

IndieanaJones committed:
Update the spider file

IndieanaJones committed:
Update the Eggs

IndieanaJones committed:
Update the stupid wizard spell nobody uses

IndieanaJones commented:
Updated, changes:

- Reverted red-eyed spider speed nerf
- Made purple-eyes (hunter) spiders slightly faster than humans as opposed to slightly slower
-  Revert burn damage change in lieu of the health nerfs being able to cover making the spiders easier to kill by themselves
- Made the flesh spiders explicitly hostile to any non-spider life, including changelings
- Nurse spiders have medhud
- Mention enriched in the one ability description
- Made a viper subtype for wizard, more health + ventcrawl
- spider types have unique names (EX hunter spider instead of giant spider)
- Spiders are affected by stamina attacks, and will slow down based on stamina damage.  Stamina damage undoes itself at a steady rate.
- You can select what spider type you want from an egg via radial menu

IndieanaJones commented:
I'm also pretty content with where there is now, could go for a testmerge at some point.

Winterous commented:
I'm Commander Shepard and this is my favourite PR on the Citadel.

IndieanaJones committed:
Fixes Stamina Not Accumilating, Makes Tarantula Immune to Stamina Damage

RuneQuester commented:
No merge yet. this is so sad

AnturK commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -589,3 +589,40 @@
+	var/spawncount = 2
+
+/datum/dynamic_ruleset/midround/spiders/execute()
+	var/list/spawn_locs = list()
```
This and code in the event is identical, reduce copypasta please.

IndieanaJones replied:
This is generally how it works for every single midround antagonist in the dynamic ruleset definitions.  I'm not saying its the right thing to be doing, but I will ask what you'd want me to do as an alternative.

AnturK commented on code/modules/mob/living/simple_animal/hostile/giant_spider.dm:
```diff
@@ -29,48 +43,51 @@
-			humanize_spider(ghost)
+/mob/living/simple_animal/hostile/poison/giant_spider/Life(mapload)
+	. = ..()
+	staminaloss = max(0, staminaloss - 10)
```
Really not a fan of implementing stamina processing as a spider snowflake here.

IndieanaJones replied:
What are your thoughts on it, then?  I'm not keen on adding stamina to all simple mobs, as that'd be something for a different PR I'd say.

IndieanaJones committed:
Merge branch 'master' into spider_rework

IndieanaJones committed:
Update Event Code

IndieanaJones committed:
Less Copypasta

IndieanaJones committed:
Move stamina down a few levels

IndieanaJones committed:
Merge branch 'master' into spider_rework

IndieanaJones committed:
Fick

IndieanaJones committed:
Merge branch 'master' into spider_rework

IndieanaJones committed:
Fix Up Simplemob Stamina

IndieanaJones committed:
Adjust speed only when necessary, and right when it happens

IndieanaJones committed:
Accept the Forced Argument
